### PR TITLE
fix(KAA): fix KAA

### PR DESCRIPTION
**Pull request by TobyBridle**
This might take a little while since I have a very important project due but shouldn't be too long.
I have added the Sapphire Duck extractor as a I'm pretty certain that it's ready but I'll be able to tell once I have the rest of the code working.

### CODE REVIEW

**Comment by TobyBridle** [comment]
I'm going to **try** and hold off on working on this for a bit, as it seems that KAA are in the process of upgrading their API.
They've just updated their `episodes` route and added a seasons route which I have accounted for but a lot of things are still returning empty objects.

File: app/src/main/java/ani/saikou/parsers/AnimeSources.kt [hunk 1]
```diff
@@ -12,7 +12,7 @@ object AnimeSources : WatchSources() {
         "Marin" to ::Marin,
         "AnimePahe" to ::AnimePahe,
         "ConsumeBili" to ::ConsumeBili,
-//        "KickAssAnime" to ::KickAssAnime
+        "KickAssAnime" to ::KickAssAnime
```
Comment (scrazzz) [inline app/src/main/java/ani/saikou/parsers/AnimeSources.kt:15]:
```suggestion
        "KickAssAnime" to ::KickAssAnime,
```

Comment (scrazzz) [inline app/src/main/java/ani/saikou/parsers/AnimeSources.kt:15]:
```suggestion
        "KickAssAnime" to ::KickAssAnime,
```

File: app/src/main/java/ani/saikou/parsers/anime/KickAssAnime.kt [hunk 2]
```diff
@@ -4,317 +4,246 @@ import android.net.Uri
 import ani.saikou.FileUrl
 import ani.saikou.Mapper
 import ani.saikou.client
+import ani.saikou.levenshtein
 import ani.saikou.parsers.*
 import ani.saikou.parsers.anime.extractors.*
+import com.lagradost.nicehttp.JsonAsString
```
Comment (scrazzz) [inline app/src/main/java/ani/saikou/parsers/anime/KickAssAnime.kt:10]: We don't use lagradost's nicehttp anymore, it was removed on [this commit](https://github.com/saikou-app/saikou/commit/4a3fcc3b36363ac0d994e83d609266b078c4e9ce)

File: app/src/main/java/ani/saikou/parsers/anime/KickAssAnime.kt [hunk 3]
```diff
@@ -4,317 +4,246 @@ import android.net.Uri
...
-        listOf(
-            json.episode.link1,
-            json.episode.link2,
-            json.episode.link3,
-            json.episode.link4
+  override val name: String = "KickAssAnime"
+  override val saveName: String = "kick_ass_anime"
+  override val hostUrl: String = "https://www2.kickassanime.ro"
+  val apiUrl = "$hostUrl/api"
+  val thumbnailUrl = "$hostUrl/images/thumbnail"
```
Comment (scrazzz) [inline app/src/main/java/ani/saikou/parsers/anime/KickAssAnime.kt:20]: These `val`s could be made private.

**Review by scrazzz — COMMENTED** [review]
Everything else looks good. Just needs some minor changes.

Ask Shivam about the nicehttp `JsonAsString()` method. I don't see that method on Shivam's new nicehttp library.

**Review by scrazzz — CHANGES_REQUESTED** [review]

File: app/src/main/java/ani/saikou/parsers/anime/KickAssAnime.kt [hunk 4]
```diff
@@ -4,317 +4,285 @@ import android.net.Uri
...
+    val json =
+            client.get(animeLink, headers = mapOf("accept" to "application/json"))
+                .parsed<EpisodesJSON>()
+
+    val episodes = mutableListOf(json.episodes)
+
+      // Smaller page number == more recent episodes
+    for (page in 1..(json.pages.toIntOrNull() ?: 1)) {
+      val url = animeLink.substringBefore("&page")
+        val pageEpisodes = client.get("$url&page=$page", headers = mapOf("accept" to "application/json")).parsed<EpisodesJSON>().episodes;
```
Comment (scrazzz) [inline app/src/main/java/ani/saikou/parsers/anime/KickAssAnime.kt:36]: Same here

**Review by scrazzz — APPROVED** [review]

**Pull request merged**